Fix a 32-bit specific bug in `fgMorphCast`

SingleAccretion opened the pull request:
It used to be that in the importer, whether the cast was to be marked as GTF_UNSIGNED was decided exclusively based on the incoming opcode. However, the flag only makes sense for casts from integral sources, and it turns out morph had a bug where it failed to clear this flag which resulted in
bad codegen.

The bug went as follows: `gtMorphCast` turns casts from an FP type into a small integer into a chain of casts: `CAST(small integer <- FP)` => `CAST(small integer <- CAST(TYP_INT <- FP))`. On 32 bit platforms, the code failed to clear the `GTF_UNSIGNED` flag from the original tree, which meant that the outer cast thought it had `TYP_UINT` as the source. This matters for checked casts: `conv.ovf.i2.un(-2.0d)`, which is a legitimate conversion, was interpreted
wrongly as an overflowing one as the resulting codegen only checked the upper bound via an unsigned compare.

The fix is two-fold: clear `GTF_UNSIGNED` for `GT_CAST` nodes with FP sources on creation and unify the 64 bit and 32 bit paths in `gtMorphCast`, which, after the removal of `GTF_UNSIGNED`, are identical.

This is a zero-diff change across all SPMI collections for Windows x64, Linux x64, Linux ARM64.

This **is not** a zero-diff change for Windows x86. Instances of bad codegen have been corrected in some tests.

Diffs for the aforementioned tests:
```
Top method regressions (bytes):
          12 (13.64% of base) : 235214.dasm - ConvTest:test25():short
          12 (13.64% of base) : 235147.dasm - ConvTest:test25():short
          10 (11.63% of base) : 235213.dasm - ConvTest:test24():byte
          10 (11.63% of base) : 235146.dasm - ConvTest:test24():byte
           9 ( 4.46% of base) : 86456.dasm - DevDiv_578217.Program:ILGEN_METHOD(ushort,float,int,ushort,ubyte,short,int):short
           7 ( 2.27% of base) : 86411.dasm - ILGEN_CLASS:ILGEN_METHOD(int):ubyte
           7 (10.45% of base) : 235247.dasm - ConvTest:test25():short
           7 (10.45% of base) : 235181.dasm - ConvTest:test25():short
           5 ( 7.69% of base) : 235246.dasm - ConvTest:test24():byte
           5 ( 2.67% of base) : 84471.dasm - FullProof:Test():int
           5 ( 7.69% of base) : 235180.dasm - ConvTest:test24():byte

Top method improvements (bytes):
         -67 (-33.00% of base) : 86409.dasm - ILGEN_CLASS:ILGEN_METHOD(short,ushort,ushort,int):float
```

`ConvTest` (`JIT\Methodical\NaN\r8nanconv_il_r\r8nanconv`), `FullProof` (`JIT\Regression\VS-ia64-JIT\V2.0-Beta2\b309576\bug2`) and `DevDiv_578217` diffs are as expected, ~~still investigating what happened in the `ILGEN` method - it looks like `UMOD` was not transfomed into `CORINFO_HELP_ULMOD`~~ - weird code, see below.

SingleAccretion committed:
Added a test

Verifying that checked '.un' casts from floats to small types are not treated as casts from unsigned types.

SingleAccretion committed:
Do not mark casts from FP types with GTF_UNSIGNED

It used to be that in the importer, whether the cast was to
be marked as GTF_UNSIGNED was decided exclusively based on
the incoming opcode. However, the flag only makes sense
for casts from integral sources, and it turns out morph
had a bug where it failed to clear this flag which resulted in
bad codegen.

The bug went as follows: "gtMorphCast" turns casts from an FP
type to a small integer into a chain of casts:
CAST(small integer <- FP) => CAST(small integer <- CAST(TYP_INT <- FP)).
On 32 bit platforms, the code failed to clear the GTF_UNSIGNED flag
from the original tree, which meant that the outer cast thought it
had TYP_UINT as the source. This matters for checked casts:
conv.ovf.i2.un(-2.0d), which is a legitimate conversion, was interpreted
wrongly as an overflowing one as the resulting codegen only checked the
upper bound via an unsigned compare.

The fix is two-fold: clear GTF_UNSIGNED for GT_CAST nodes with FP sources
on creation and unify the 64 bit and 32 bit paths in "gtMorphCast",
which, after the removal of GTF_UNSIGNED handling, are identical.

This is a zero-diff change across all SPMI collections for Windows x64, Linux x64,
Linux ARM64.

This **is not** a zero-diff change for Windows x86.
Instances of bad codegen have been corrected in some tests.

SingleAccretion commented:
So, the `UMOD` case is weird because the code that is responsible for it is questionable:
https://github.com/dotnet/runtime/blob/5a0edac8a817f6974d58d4f9ea8e6809796f2853/src/coreclr/jit/morph.cpp#L12618-L12662

It appears that the intention was to check whether the operands of `UMOD` are unsigned, but that doesn't make a lot of sense because `GTF_UNSIGNED` is not responsible for that, and the fact that the node is a `UMOD` is already enough - it treats its operands as unsigned. In any case, the diff is caused by this "do `GTF_UNSIGNED`s match" logic.
Base:
```scala
Before:
               [000028] ------------                            +--*  UMOD      long  
               [000025] ---------U--                            |  +--*  CAST      long <- ulong <- double
               [000024] ------------                            |  |  \--*  NEG       double
               [000023] ------------                            |  |     \--*  LCL_VAR   double V11 loc7         
               [000027] ------------                            |  \--*  CNS_LNG   long   0x000000002487c9b5
                                                                
After:                                                          
               [000028] --CX-+------                            +--*  CALL help long   HELPER.CORINFO_HELP_ULMOD
               [000025] --C--+---U-- arg0                       |  +--*  CALL help long   HELPER.CORINFO_HELP_DBL2ULNG
               [000024] -----+------ arg0                       |  |  \--*  NEG       double
               [000023] -----+------                            |  |     \--*  LCL_VAR   double V11 loc7         
               [000027] -----+------ arg1                       |  \--*  CNS_LNG   long   0x000000002487c9b5
```
Diff:
```scala
Before:
               [000028] ------------                            +--*  UMOD      long  
               [000025] ------------                            |  +--*  CAST      long <- ulong <- double
               [000024] ------------                            |  |  \--*  NEG       double
               [000023] ------------                            |  |     \--*  LCL_VAR   double V11 loc7         
               [000027] ------------                            |  \--*  CNS_LNG   long   0x000000002487c9b5
                                                                
After:                                                          
               [000028] --C--+------                            +--*  UMOD      long  
               [000025] --C--+------                            |  +--*  CALL help long   HELPER.CORINFO_HELP_DBL2ULNG
               [000024] -----+------ arg0                       |  |  \--*  NEG       double
               [000023] -----+------                            |  |     \--*  LCL_VAR   double V11 loc7         
               [000027] ------------                            |  \--*  CNS_LNG   long   0x000000002487c9b5
```
While it is weird, it should be correct.

Another note on the weird `UMOD` code: it is testing for a cast from a constant - that may be a remnant of the fact `ldc.i4` `conv.i8` used not to be folded.

SingleAccretion commented on src/tests/issues.targets:
```diff
@@ -1310,6 +1310,9 @@
             <Issue>needs triage</Issue>
         </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/JIT/Directed/Convert/signed_overflow_conversions_are_not_treated_as_unsigned/**">
+            <Issue>https://github.com/dotnet/runtime/issues/51323</Issue>
```
Right now the issue refers to a non-existnet (reverted) test. I will update it to refer to this test if/when this is merged (in the case of this not being merged, I will close it).

SingleAccretion commented:
Ready for review.

cc @sandreenko for another morph change...

sandreenko commented:
Thanks, @SingleAccretion, how did you find this issue? 

It is great that you have added a new test for such cases.

I will try to take a look this week, but such change would require a second approve, I think, PTAL @dotnet/jit-contrib

SingleAccretion commented:
> how did you find this issue?

From experimenting with `fgMorphCast`, I suppose. I noticed that these two branches were doing almost the same thing, made them do the same thing, and while inspecting the diffs realized the (previous) codegen was wrong.

AndyAyersMS reviewed (approved):
LGTM

sandreenko commented on src/coreclr/jit/compiler.hpp:
```diff
@@ -1410,10 +1410,20 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
 {
-    GenTreeCast* res = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
-    return res;
+    GenTreeCast* cast = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
```
we are calling GenTreeCast constructor:
https://github.com/dotnet/runtime/blob/419765391c14cb12366ee111b16e91a481ef7bc7/src/coreclr/jit/gentree.h#L3531-L3535
and after that we call `CanonicalizeCastNode` to clear the flag in some cases, is it worth it?
One option is to do 
```
inline GenTreeCast* Compiler::gtNewCastNode(var_types typ, GenTree* op1, bool fromUnsigned, var_types castType)
{
  if (varTypeIsFloating(cast->CastOp())) { fromUnsigned = false; }
  GenTreeCast* cast = new (this, GT_CAST) GenTreeCast(typ, op1, fromUnsigned, castType);
```

or add this logic to `GenTreeCast` constructor (note it will affect more cases then).

Often I set conditional breakpoints on flag values so I would prefer not to change it back and forth without a reason.

Have you tried to add `assert(!varTypeIsFloating(cast->CastOp() || !fromUnsigned)` and fix the callers who call it with wrong value of `bool fromUnsigned`?

SingleAccretion replied:
> Have you tried to add `assert(!varTypeIsFloating(cast->CastOp() || !fromUnsigned)` and fix the callers who call it with wrong value of bool `fromUnsigned`?

Yes, in fact that was the original version of the fix, and the only place I had to fix was the importer. I went back and forth on this a few times, and ultimately decided to have the change initially be with this more conservative implementation, because it is "guaranteed" (not really because I did not add it to the constructor, but nothing calls the constructor directly) to always return "correct" result to the caller. I suppose I should revisit it and run the `assert`ty version through the CI. Maybe it turns out more callers are getting it wrong and it would be better to normalize centrally (regardless of that, the flag flipping will be fixed).

sandreenko replied:
> I suppose I should revisit it and run the assertty version through the CI. Maybe it turns out more callers are getting it wrong and it would be better to normalize centrally (regardless of that, the flag flipping will be fixed).

I would prefer this if you have time, if not please open a follow-up issue to investigate/clean it and we will merge this PR as is.

SingleAccretion replied:
> I would prefer this

I would as well, so we'll investigate it as part of this PR.

SingleAccretion committed:
Assert instead of normalizing

Instead of normalizing GTF_UNSIGNED for FP sources in "gtNewCastNode",
assert that it is not set in GenTreeCast's constructor and fix the
importer to respect that constraint.

sandreenko reviewed (approved):
LGTM, thank you!
